packages/core/echo/echo-pipeline/src/automerge/mesh-echo-replicator-connection.ts

```diff
@@ -0,0 +1,130 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { cbor, type Message } from '@dxos/automerge/automerge-repo';
+import { Resource } from '@dxos/context';
+import { invariant } from '@dxos/invariant';
+import { type PublicKey } from '@dxos/keys';
+import { log } from '@dxos/log';
+import { AutomergeReplicator, type AutomergeReplicatorFactory } from '@dxos/teleport-extension-automerge-replicator';
+
+import type { ReplicatorConnection, ShouldAdvertizeParams } from './echo-replicator';
+
+const DEFAULT_FACTORY: AutomergeReplicatorFactory = (params) => new AutomergeReplicator(...params);
+
+export type MeshReplicatorConnectionParams = {
+  ownPeerId: string;
+  onRemoteConnected: () => void;
+  onRemoteDisconnected: () => void;
+  shouldAdvertize: (params: ShouldAdvertizeParams) => Promise<boolean>;
+  replicatorFactory?: AutomergeReplicatorFactory;
+};
+
+export class MeshReplicatorConnection extends Resource implements ReplicatorConnection {
+  public readable: ReadableStream<Message>;
+  public writable: WritableStream<Message>;
+  public remoteDeviceKey: PublicKey | null = null;
+
+  public readonly replicatorExtension: AutomergeReplicator;
+
+  private _remotePeerId: string | null = null;
+  private _isEnabled = false;
+
+  constructor(private readonly _params: MeshReplicatorConnectionParams) {
+    super();
+
+    let readableStreamController!: ReadableStreamDefaultController<Message>;
+    this.readable = new ReadableStream<Message>({
+      start: (controller) => {
+        readableStreamController = controller;
+        this._ctx.onDispose(() => controller.close());
+      },
+    });
+
+    this.writable = new WritableStream<Message>({
+      write: async (message: Message, controller) => {
+        invariant(this._isEnabled, 'Writing to a disabled connection');
+        try {
+          await this.replicatorExtension.sendSyncMessage({ payload: cbor.encode(message) });
```

Comment:
does this mean that we would only be able to have 1 outgoing sync message in-transit for every space?

Reply:
1 outgoing sync message per peer